src/core/ComboBox/ComboBox.tsx

```diff
@@ -380,84 +459,365 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        onFocus={handleFocus}
+        aria-activedescendant={
+          isOpen && focusedIndex != undefined && focusedIndex > -1
+            ? getIdFromIndex(focusedIndex)
+            : undefined
+        }
+        role='combobox'
+        aria-controls={isOpen ? `${id}-list` : undefined}
+        aria-autocomplete='list'
+        spellCheck={false}
+        autoCapitalize='none'
+        autoCorrect='off'
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxInput.displayName = 'ComboBoxInput';
+
+const ComboBoxPopover = React.forwardRef(
+  (
+    props: PopoverProps & { children: JSX.Element },
+    forwardedRef: React.RefObject<Element>,
+  ) => {
+    const { children, ...rest } = props;
+    const { isOpen } = useSafeContext(ComboBoxStateContext);
+    const dispatch = useSafeContext(ComboBoxActionContext);
+    const { inputRef, toggleButtonRef } = useSafeContext(ComboBoxRefsContext);
+
+    // sync internal isOpen state with user's visible prop
+    React.useEffect(() => {
+      dispatch([props.visible ? 'open' : 'close']);
+    }, [dispatch, props.visible]);
+
+    return (
+      <Popover
+        placement='bottom-start'
+        visible={isOpen}
+        onClickOutside={React.useCallback(
+          (_, { target }) => {
+            if (!toggleButtonRef.current?.contains(target as Element)) {
+              dispatch(['close']);
             }
-          }}
+          },
+          [dispatch, toggleButtonRef],
+        )}
+        animation='shift-away'
+        duration={200}
+        reference={inputRef}
+        ref={forwardedRef}
+        content={children}
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxPopover.displayName = 'ComboBoxPopover';
+
+const ComboBoxMenu = React.forwardRef(
+  (
+    props: Omit<MenuProps, 'onClick'> & React.ComponentPropsWithoutRef<'ul'>,
+    forwardedRef: React.RefObject<HTMLUListElement>,
+  ) => {
+    const { className, style, ...rest } = props;
+    const { minWidth, id } = useSafeContext(ComboBoxStateContext);
+    const { menuRef, optionsRef } = useSafeContext(ComboBoxRefsContext);
+
+    React.useEffect(() => {
+      if (menuRef.current != null) {
+        const items = menuRef.current.querySelectorAll('[data-iui-index]');
+        optionsRef.current = Array.from(items) as HTMLLIElement[]; // need this cast to make TS happy :(
+      }
+    }, [menuRef, optionsRef, props.children]);
+
+    const refs = useMergedRefs(menuRef, forwardedRef);
+
+    return (
+      <Menu
+        id={`${id}-list`}
+        style={React.useMemo(
+          () => ({
+            minWidth,
+            maxWidth: `min(${minWidth * 2}px, 90vw)`,
+            maxHeight: 300,
+            ...style,
+          }),
+          [minWidth, style],
+        )}
+        setFocus={false}
+        role='listbox'
+        ref={refs}
+        className={cx('iui-scroll', className)}
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxMenu.displayName = 'ComboBoxMenu';
+
+const ComboBoxMenuItem = React.memo(
```

Comment:
Why memo?

Reply:
To avoid excessive rerenders. Is there a downside to not having it?

Reply:
oh, if you checked the performance gain, thats good.